File: cryptography/ver2/encryption.py
import base64
import hmac
import hashlib

# コンテストで共通の秘密鍵をここに置く
secret = b"MY_SECRET_KEY"

code = input()

payload, sig = code.split(".")

payload += "="*(-len(payload)%4)
sig += "=" * (-len(sig)%4)

msg = base64.urlsafe_b64decode(payload)
recv_sig = base64.urlsafe_b64decode(sig)

calc_sig = hmac.new(secret,msg,hashlib.sha256).digest()

if not hmac.compare_digest(calc_sig, recv_sig):
    print("改ざんされています")
    exit()

dt,alphabet,P = msg.decode().split(":")

print(f"問題: {alphabet} \n 経過時間: {dt} \n 得点: {P}")

File: cryptography/ver2/encryption.js
// この問題のアルファベットを入れてください
  const alphabet = "A";

  // この問題の得点を入れてください
  const points = [50,125];

  // 小課題ごとの対応するテストケース番号を入れてください
  const tasks = [[1],[1,2]];

  // コンテストで共通する秘密鍵を入力してください(英文字で入力、空白禁止)
  const secret = "MY_SECRET_KEY";

// 中略...

// Base64URLエンコード
  function base64urlEncode(bytes) {
    let binary = "";
    for (const b of bytes) binary += String.fromCharCode(b);

    return btoa(binary)
      .replace(/\+/g,"-")
      .replace(/\//g,"_")
      .replace(/=+$/,"");
  }

  // HMAC-SHA256
  async function hmacSHA256(message) {
    const key = await crypto.subtle.importKey(
      "raw",
      new TextEncoder().encode(secret),
      { name: "HMAC", hash: "SHA-256"},
      false,
      ["sign"]
    );

    const sig = await crypto.subtle.sign(
      "HMAC",
      key,
      new TextEncoder().encode(message)
    );

    return new Uint8Array(sig);
  }

  async function makeCode(dt,ch,P) {
    const msg = `${dt}:${ch}:${P}`;

    const payload = base64urlEncode(
      new TextEncoder().encode(msg)
    );

    const sig = base64urlEncode(
      await hmacSHA256(msg)
    );

      return payload + "." + sig;
  }
